iis-wf/iis-wf-referenceextraction/src/main/java/eu/dnetlib/iis/wf/referenceextraction/patent/PatentMetadataRetrieverJob.java

```diff
@@ -0,0 +1,167 @@
+            
+            try {
+                PatentServiceFacade patentServiceFacade = ServiceFacadeUtils.instantiate(sc.hadoopConfiguration());
+                
+                JavaRDD<ImportedPatent> importedPatents = avroLoader.loadJavaRDD(sc, params.inputPath,
+                        ImportedPatent.class);
+
+                Tuple2<JavaRDD<DocumentText>, JavaRDD<Fault>> results = obtainMetadata(importedPatents, patentServiceFacade);
+                
+                storeInOutput(results._1, results._2, generateReportEntries(sc, results._1, results._2), outputPaths,
+                        params.numberOfEmittedFiles);
+                
+            } catch (ServiceFacadeException e) {
+                throw new RuntimeException("unable to instantiate patent service facade!", e);
+            }
+        }
+    }
+
+    //------------------------ PRIVATE --------------------------
+
+    private static JavaRDD<ReportEntry> generateReportEntries(JavaSparkContext sparkContext, 
+            JavaRDD<DocumentText> processedEntities, JavaRDD<Fault> processedFaults) {
+        
+        Preconditions.checkNotNull(sparkContext, "sparkContext has not been set");
+        
+        ReportEntry processedEnttiesCounter = ReportEntryFactory.createCounterReportEntry(COUNTER_PROCESSED_TOTAL, processedEntities.count());
+        ReportEntry processedFaultsCounter = ReportEntryFactory.createCounterReportEntry(COUNTER_PROCESSED_FAULT, processedFaults.count());
+        
+        return sparkContext.parallelize(Lists.newArrayList(processedEnttiesCounter, processedFaultsCounter));
+    }
+    
+    private static void storeInOutput(JavaRDD<DocumentText> retrievedPatentMeta, 
+            JavaRDD<Fault> faults, JavaRDD<ReportEntry> reports, OutputPaths outputPaths, int numberOfEmittedFiles) {
+        avroSaver.saveJavaRDD(retrievedPatentMeta.coalesce(numberOfEmittedFiles), DocumentText.SCHEMA$, outputPaths.getResult());
+        avroSaver.saveJavaRDD(faults.coalesce(numberOfEmittedFiles), Fault.SCHEMA$, outputPaths.getFault());
+        avroSaver.saveJavaRDD(reports.coalesce(numberOfEmittedFiles), ReportEntry.SCHEMA$, outputPaths.getReport());
+    }
+    
+    private static Tuple2<JavaRDD<DocumentText>, JavaRDD<Fault>> obtainMetadata(JavaRDD<ImportedPatent> input, PatentServiceFacade patentServiceFacade) {
+        JavaPairRDD<CharSequence, ContentRetrieverResponse> obtainedIdToResponce = input
+                .mapToPair(x -> new Tuple2<CharSequence, ContentRetrieverResponse>(getId(x),
+                        obtainMetadata(x, patentServiceFacade)));
+        obtainedIdToResponce.persist(StorageLevel.DISK_ONLY());
+        return new Tuple2<>(
+                obtainedIdToResponce.map(e -> DocumentText.newBuilder().setId(e._1).setText(e._2.getContent()).build()),
+                obtainedIdToResponce.filter(e -> e._2.getException() != null).map(e -> FaultUtils.exceptionToFault(e._1, e._2.getException(), null)));
+    }
```

Comment:
Yes, you are right. This code begged for refactoring. I've removed this method because it hardly helped in understanding what is going on. Not it should be more straightforward.

Reply:
I'd still create two methods encapsulating RDD processing for retrieved patent metadata and faults, just to make the code more readable.

Reply:
I can see this was already committed in https://github.com/openaire/iis/pull/1091/commits/e3c3d1e30abffdff3081aa033b69b0e3613ad21e#diff-32e05d485e9616c754f5c2f56468ddb8. Both `JavaRDD<DocumentText>` and `JavaRDD<Fault>` RDDs are produced separately.